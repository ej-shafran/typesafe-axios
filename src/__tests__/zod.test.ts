import { taxios } from "..";

import { z } from "zod";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = z.object({
  id: z.string(),
  username: z.string(),
  email: z.string(),
  address: z.object({
    street: z.string(),
    zipcode: z.string().optional(),
  }),
  followers: z.array(
    z.object({
      id: z.string(),
      role: z.enum(["admin", "guest"]).optional(),
    })
  ),
});

type TYPE = z.infer<typeof SCHEMA>;
function ensure(_: TYPE) { }

describe("zod interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);
    const { data } = await taxios(SCHEMA).get(UTILS.API);

    expect(data).toEqual(UTILS.VALID);
    ensure(data);
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    try {
      await taxios(SCHEMA).get(UTILS.API);
      fail("Did not throw error.");
    } catch (error) {
      expect(JSON.stringify(error, null, 2)).toMatchSnapshot();
    }
  });
});
