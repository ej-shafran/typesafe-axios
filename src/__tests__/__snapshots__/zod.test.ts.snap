// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zod interop throws a validation error when the data is invalid 1`] = `
"{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "email"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "object",
      "received": "string",
      "path": [
        "address"
      ],
      "message": "Expected object, received string"
    }
  ],
  "name": "ZodError"
}"
`;
